File: generated_code/GeminiPro/Python/096_SplayTree.py
class SplayTree:
    """
    A self-contained Python class implementing a Splay Tree.

    This data structure acts as a dictionary-like set for storing unique integers.
    It supports `insert`, `delete`, and `search` operations. The defining
    feature of a Splay Tree is its self-balancing mechanism: whenever a key
    is accessed, it is moved to the root of the tree through a series of
    rotations. This "splaying" operation makes frequently accessed elements
    faster to retrieve over time.
    """

    class _Node:
        """A private node class for the Splay Tree."""
        __slots__ = 'key', 'parent', 'left', 'right'

        def __init__(self, key: int, parent: 'SplayTree._Node' = None,
                     left: 'SplayTree._Node' = None, right: 'SplayTree._Node' = None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root: SplayTree._Node | None = None

    def _rotate_left(self, x: _Node):
        """Performs a left rotation on the given node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _rotate_right(self, x: _Node):
        """Performs a right rotation on the given node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node: _Node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        if not node:
            return
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if node == parent.left:
                    self._rotate_right(parent)
                else:
                    self._rotate_left(parent)
            elif node == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._rotate_right(grandparent)
                self._rotate_right(parent)
            elif node == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._rotate_left(grandparent)
                self._rotate_left(parent)
            elif node == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._rotate_left(parent)
                self._rotate_right(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._rotate_right(parent)
                self._rotate_left(grandparent)
        self.root = node

    def search(self, key: int) -> bool:
        """
        Searches for a key in the tree and splays the accessed node.

        If the key is found, the node containing the key is splayed to the
        root and the method returns True.
        If the key is not found, the last node visited before the search
        terminated is splayed to the root, and the method returns False.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found
                self._splay(current)
                return True

        # Key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
        return False

    def insert(self, key: int):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted and then splayed to the root.

        Args:
            key: The integer key to insert.
        """
        # Case 1: The tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree on the key. The root will be the closest node.
        # This conveniently brings the relevant part of the tree to the top.
        self.search(key)

        # After splaying, check if the key is already at the root.
        if self.root.key == key:
            return  # Key already exists

        # Case 2: Insert the new node and restructure the tree
        new_node = self._Node(key)
        if key < self.root.key:
            # New node becomes the root.
            # Old root becomes the right child of the new node.
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
            self.root.parent = new_node
        else:  # key > self.root.key
            # New node becomes the root.
            # Old root becomes the left child of the new node.
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None
            self.root.parent = new_node

        self.root = new_node

    def delete(self, key: int):
        """
        Deletes a key from the tree.

        The tree is first splayed on the key. If the key exists, it is
        removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        # Splay the node to be deleted (or its parent) to the root.
        if not self.search(key):
            # Key was not found, so nothing to delete.
            # The search call already splayed the closest node.
            return

        # At this point, the node with the key is guaranteed to be the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left child, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Disconnect the left subtree to work on it.
            left_subtree.parent = None

            # Find the maximum node in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this maximum node to the root of the (temporary) left subtree.
            # We can do this by temporarily setting self.root.
            self.root = left_subtree
            self._splay(max_node)

            # After splaying, self.root is the max_node. It has no right child.
            # We can now attach the original right subtree to it.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/096_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree.
 * The Splay Tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * This implementation stores integer keys and functions as a set (no duplicates).
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. Brings the node with the given key
     * (or the last accessed node if the key is not found) to the root.
     * This is a recursive, bottom-up splay implementation.
     * @param root The root of the tree/subtree to splay.
     * @param key The key to splay towards the root.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node root, int key) {
        // Base case: root is null or key is present at root
        if (root == null || root.key == key) {
            return root;
        }

        // Key lies in the left subtree
        if (key < root.key) {
            // Key is not in the tree, we are done
            if (root.left == null) {
                return root;
            }

            // Zig-Zig (Left Left)
            if (key < root.left.key) {
                // Recursively bring the key as root of left-left
                root.left.left = splay(root.left.left, key);
                // Do first rotation for root
                root = rightRotate(root);
            }
            // Zig-Zag (Left Right)
            else if (key > root.left.key) {
                // Recursively bring the key as root of left-right
                root.left.right = splay(root.left.right, key);
                // Do first rotation for root.left
                if (root.left.right != null) {
                    root.left = leftRotate(root.left);
                }
            }

            // Do second rotation for root (or the single rotation in Zig case)
            return (root.left == null) ? root : rightRotate(root);
        }
        // Key lies in the right subtree
        else { // key > root.key
            // Key is not in the tree, we are done
            if (root.right == null) {
                return root;
            }

            // Zig-Zag (Right Left)
            if (key < root.right.key) {
                // Recursively bring the key as root of right-left
                root.right.left = splay(root.right.left, key);
                // Do first rotation for root.right
                if (root.right.left != null) {
                    root.right = rightRotate(root.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > root.right.key) {
                // Recursively bring the key as root of right-right
                root.right.right = splay(root.right.right, key);
                // Do first rotation for root
                root = leftRotate(root);
            }

            // Do second rotation for root (or the single rotation in Zig case)
            return (root.right == null) ? root : leftRotate(root);
        }
    }

    // --- Public API ---

    /**
     * Inserts a key into the Splay Tree. If the key already exists,
     * the tree remains unchanged. After insertion, the new node becomes the root.
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // The new node becomes the new root. The old root and one of its subtrees
        // become the children of the new node. This is a "split" operation.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the Splay Tree.
     * This operation splays the found node (or the last accessed node) to the root.
     * @param key The integer key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Deletes a key from the Splay Tree. If the key is not found,
     * the tree is splayed based on the last accessed node and remains otherwise unchanged.
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        // If the tree is empty
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // Key is at the root, now we need to join its two subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to the top.
            // Since `key` is greater than any element in the left subtree, splaying
            // for `key` will bring the max element to the root of the left subtree.
            Node newRoot = splay(leftSubtree, key);
            
            // The new root (max of left subtree) will have no right child.
            // Attach the original right subtree as the right child of the new root.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}